File: leet-code/2021/0011-container-with-most-water.rs
// https://leetcode.com/problems/container-with-most-water/

impl Solution {
    pub fn max_area(heights: Vec<i32>) -> i32 {
        let (mut start, mut end) = (0usize, heights.len() -1);
        let mut max_area: i32 = 0;

        while start < end {
            let min_height = std::cmp::min(heights[start], heights[end]);
            max_area = std::cmp::max(max_area, min_height * (end-start) as i32);
            match heights[start] < heights[end] {
                true  => start += 1,
                false => end -= 1
            }
        }
        max_area
    }
}

impl OptimizedSolution {
    pub fn max_area(heights: Vec<i32>) -> i32 {
        let (mut start, mut end) = (0usize, heights.len() -1);
        let mut max_area: i32 = 0;

        while start < end {
            let min_height = std::cmp::min(heights[start], heights[end]);
            max_area = std::cmp::max(max_area, min_height*(end-start) as i32);
            match heights[start] < heights[end] {
                true  => {
                    // optimization. does not make any difference in LC results though
                    while start + 1 < end && heights[start] >= heights[start+1] {
                        start += 1;
                    }
                    start += 1;
                },
                false => {
                    // optimization. does not make any difference in LC results though
                    while start + 1 < end && heights[end] >= heights[end-1] {
                        end -= 1;
                    }
                    end -= 1;
                }
            }
        }
        max_area
    }
}

File: leet-code/2022/0011-container-with-most-water.cpp
// https://leetcode.com/problems/container-with-most-water/
// Medium

class Solution {
public:
    int maxArea(vector<int>& height) {
        int max = 0;
        for (int i=0, j=height.size()-1; i < j; ) {
            max = std::max(max, std::min(height[i], height[j])*(j-i));
            bool moveLeft = height[i] < height[j];
            i += moveLeft;
            j -= !moveLeft;
        }
        return max;
    }
};

class LongerSolution {
public:
    int maxArea(vector<int>& height) {
        int max = 0;
        for (int i = 0, j = height.size()-1; i < j; ) {
            int amount = std::min(height[j], height[i]) * (j-i);
            max = std::max(amount, max);
            if (height[i] < height[j]) {
                ++i;
            } else {
                --j;
            }
        }
        return max;
    }
};

class BranchlessSolution {
public:
    int maxArea(vector<int>& height) {
        int max = 0;
        for (int i = 0, j = height.size()-1; i < j; ) {
            int amount = std::min(height[j], height[i]) * (j-i);
            max = std::max(amount, max);
            bool index = bool(height[i] < height[j]);
            i = std::array<int, 2>{i, i+1}[index];
            j = std::array<int, 2>{j-1, j}[index];
        }
        return max;
    }
};

class OldSolution {
public:
    int maxArea(vector<int>& height) {
        int max = (height.size()-1) * std::min(height.front(), height.back());
        for (int i=0, j=height.size()-1; i < j; ) {
            max = std::max(max, (j-i) * std::min(height[i], height[j]));
            if (height[i] < height[j]) {
                ++i;
            } else {
                --j;
            }
        }
        return max;
    }
};

/*
Test data:
[1,8,6,2,5,4,8,3,7]
[1,10,2,3,4,5,10,1]
[1,1]
[10,1,1]
[1,10,10]
[1,10,1]
[10,1,10]
[1,2,4,3]
*/
